zebra-chain/src/block/serialize.rs

```diff
@@ -22,9 +22,40 @@ use crate::{
 /// transaction in the chain is approximately 1.5 kB smaller.)
 pub const MAX_BLOCK_BYTES: u64 = 2_000_000;
 
+/// Checks if a block header version is valid
+fn check_version(version: u32) -> Result<(), &'static str> {
+    match version {
+        // The Zcash specification says that:
+        // "The current and only defined block version number for Zcash is 4."
+        // but this is not actually part of the consensus rules, and in fact
+        // broken mining software created blocks that do not have version 4.
+        // There are approximately 4,000 blocks with version 536870912; this
```

Comment:
Can you add documentation to explain when Zebra could encounter invalid versions when serializing a block header?

I think these errors could only happen if we constructed a generated block in memory with the wrong version. All other blocks are deserialized when we receive them, and never modified. So the deserialisation would pick up any errors.

And in the getblocktemplate RPC, we generate a template with version 4, then the miner generates the actual block, and then we deserialize it and do this check.

Can we also add one of the high version blocks to our test vectors, and add a test that serializes it? That will make sure we don't panic anywhere in that code path.

Reply:
> Can you add documentation to explain when Zebra could encounter invalid versions when serializing a block header?
> 
> I think these errors could only happen if we constructed a generated block in memory with the wrong version. All other blocks are deserialized when we receive them, and never modified. So the deserialisation would pick up any errors.
> 
> And in the getblocktemplate RPC, we generate a template with version 4, then the miner generates the actual block, and then we deserialize it and do this check.

Yes, added in https://github.com/ZcashFoundation/zebra/pull/6475/commits/7d7594f86b9ae0dab79a2e41792b7f41c89a6a33.

Reply:
Added testing for block header (de)serialization errors in https://github.com/ZcashFoundation/zebra/pull/6475/commits/6b4ec1b79804e0bd9c7b3f61f3c060aa44a49b8c.